src/NHibernate/Linq/Visitors/WhereJoinDetector.cs

```diff
@@ -92,7 +93,23 @@ public void Transform(WhereClause whereClause)
 			}
 		}
 
-		public override Expression VisitExpression(Expression expression)
+		public void Transform(NhHavingClause whereClause)
+		{
+			whereClause.TransformExpressions(Visit);
+
+			var values = _values.Pop();
+
+			foreach (var memberExpression in values.MemberExpressions)
+			{
+				// If outer join can never produce true, we can safely inner join.
+				if (!values.GetValues(memberExpression).Contains(true))
+				{
+					_joiner.MakeInnerIfJoined(memberExpression);
+				}
+			}
```

Comment:
From [L100](https://github.com/nhibernate/nhibernate-core/pull/568/files#diff-4106d8fef553f297317a11b60b40b815R100) till here, could be refactored in a method with code from [L84](https://github.com/nhibernate/nhibernate-core/pull/568/files#diff-4106d8fef553f297317a11b60b40b815R84) to [L93](https://github.com/nhibernate/nhibernate-core/pull/568/files#diff-4106d8fef553f297317a11b60b40b815R93).